src/enforcers/ERC721TransferEnforcer.sol

```diff
@@ -0,0 +1,57 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.23;
+
+import { CaveatEnforcer } from "./CaveatEnforcer.sol";
+import { ModeCode } from "../utils/Types.sol";
+import { IERC721 } from "@openzeppelin/contracts/token/ERC721/IERC721.sol";
+import { ExecutionLib } from "@erc7579/lib/ExecutionLib.sol";
+
+contract ERC721TransferEnforcer is CaveatEnforcer {
+    error UnauthorizedTransfer();
+
+    function beforeHook(
+        bytes calldata _terms,
+        bytes calldata,
+        ModeCode,
+        bytes calldata executionCallData,
+        bytes32,
+        address,
+        address
+    )
+        public
+        virtual
+        override
+    {
+        (address permittedContract, uint256 permittedTokenId) = getTermsInfo(_terms);
+        (address target,, bytes calldata callData_) = ExecutionLib.decodeSingle(executionCallData);
```

Comment:
We are assuming this is a single execution, we can use this modifier `onlySingleExecutionMode`